File: Tests/Tests/STPTelemetryClientTest.m
//
//  STPTelemetryClientTest.m
//  Stripe
//
//  Created by Ben Guo on 4/18/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "STPTelemetryClient.h"
#import "STPAPIClient.h"

@interface STPTelemetryClientTest : XCTestCase

@end

@implementation STPTelemetryClientTest

- (void)testAddTelemetryData {
    STPTelemetryClient *sut = [STPTelemetryClient sharedInstance];
    NSMutableDictionary *params = [@{@"foo": @"bar"} mutableCopy];
    XCTestExpectation *exp = [self expectationWithDescription:@"delay"];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1*NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [sut addTelemetryFieldsToParams:params];
        XCTAssertNotNil(params[@"muid"]);
        [exp fulfill];
    });
    [self waitForExpectationsWithTimeout:2 handler:nil];
}

- (void)testAdvancedFraudSignalsSwitch {
    XCTAssertTrue([Stripe advancedFraudSignalsEnabled]);
    [Stripe setAdvancedFraudSignalsEnabled:NO];
    XCTAssertFalse([Stripe advancedFraudSignalsEnabled]);
}

@end

File: Tests/Tests/STPTelemetryClientTest.swift
//
//  STPTelemetryClientTest.swift
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 9/24/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import XCTest

@testable import Stripe

class STPTelemetryClientTest: XCTestCase {

  func testAddTelemetryData() {
    let sut = STPTelemetryClient.sharedInstance()
    var params: [String: Any] = [
      "foo": "bar"
    ]
    let exp = expectation(description: "delay")
    DispatchQueue.main.asyncAfter(
      deadline: DispatchTime.now() + Double(Int64(0.1 * Double(NSEC_PER_SEC)))
        / Double(NSEC_PER_SEC),
      execute: {
        sut.addTelemetryFields(toParams: &params)
        XCTAssertNotNil(params)
        exp.fulfill()
      })
    waitForExpectations(timeout: 2, handler: nil)
  }

  func testAdvancedFraudSignalsSwitch() {
    XCTAssertTrue(StripeAPI.advancedFraudSignalsEnabled)
    StripeAPI.advancedFraudSignalsEnabled = false
    XCTAssertFalse(StripeAPI.advancedFraudSignalsEnabled)
  }

}
